cpp/src/decisiontree/batched-levelalgo/builder_base.cuh

```diff
@@ -27,10 +30,104 @@
 #include "split.cuh"
 
 #include <common/nvtx.hpp>
+#include <deque>
+#include <utility>
 
 namespace ML {
 namespace DT {
 
+template <typename NodeT>
+class NodeQueue {
+  const DecisionTreeParams params;
+  std::vector<NodeT> tree_;
+  std::deque<NodeWorkItem> work_items_;
```

Comment:
Just an observation - seems like the information stored in `NodeWorkItem` struct is a subset of `NodeT` ; and that the main function of `NodeWorkItem` is used to keep track of information that is anyway stored in `NodeT`, but on a per-batch-basis... what about an alternative where we just keep track of the indices of `tree_` in an array? would that lead to complications that I'm maybe not accounting for?

Reply:
Yes that's exactly right. I was thinking of removing NodeT in my next PR.